framework/include/materials/MaterialProperty.h

```diff
@@ -170,123 +178,193 @@ class MaterialProperty : public PropertyValue
    */
   virtual void load(std::istream & stream) override;
 
-  /**
-   * copy the value portion (not the derivatives) of the DualNumber<Real> version of the material
-   * property to the Real version for the specified quadrature point
-   */
-  void copyDualNumberToValue(const unsigned int i) override { _value[i].synchronizeToVal(); }
-
-  void markAD(bool use_ad) override;
-
 private:
   /// private copy constructor to avoid shallow copying of material properties
-  MaterialProperty(const MaterialProperty<T> & /*src*/)
+  MaterialPropertyBase(const MaterialPropertyBase<T, is_ad> & /*src*/)
   {
     mooseError("Material properties must be assigned to references (missing '&')");
   }
 
   /// private assignment operator to avoid shallow copying of material properties
-  MaterialProperty<T> & operator=(const MaterialProperty<T> & /*rhs*/)
+  MaterialPropertyBase<T, is_ad> & operator=(const MaterialPropertyBase<T, is_ad> & /*rhs*/)
   {
     mooseError("Material properties must be assigned to references (missing '&')");
   }
 
 protected:
-  /// Whether this property was declared as AD
-  bool _use_ad;
   /// Stored parameter value.
-  std::vector<MooseADWrapper<T>> _value;
+  std::vector<MooseADWrapper<T, is_ad>> _value;
 };
 
+template <typename T>
+class MaterialProperty;
+template <typename T>
+class ADMaterialProperty;
+
 // ------------------------------------------------------------
 // Material::Property<> class inline methods
 
-template <typename T>
-inline void
-MaterialProperty<T>::markAD(bool use_ad)
+namespace MaterialPropertyPrivate
```

Comment:
I've been trying to use an `moose::internal` namespace nesting for this kind of stuff.